src/solc_0.8/StarterPack/StarterPackV2.sol

```diff
@@ -0,0 +1,382 @@
+pragma solidity 0.8.2;
+
+import "./PurchaseValidator.sol";
+import "../catalyst/GemsCatalystsRegistry.sol";
+import "../common/BaseWithStorage/ERC2771Handler.sol";
+import "../common/Libraries/SafeMathWithRequire.sol";
+
+/// @title StarterPack contract that supports SAND as payment
+/// @notice This contract manages the purchase and distribution of StarterPacks (bundles of Catalysts and Gems)
+contract StarterPackV2 is PurchaseValidator, ERC2771Handler {
+    using SafeMathWithRequire for uint256;
+    uint256 private constant DECIMAL_PLACES = 1 ether;
+
+    address internal immutable _sand;
+    address internal immutable _registry;
+    bool public _sandEnabled;
+
+    // Mapping catalyst and gem ids to their prices
+    mapping(uint16 => uint256) private _catalystPrices;
+    mapping(uint16 => uint256) private _catalystPreviousPrices;
+    mapping(uint16 => uint256) private _gemPrices;
+    mapping(uint16 => uint256) private _gemPreviousPrices;
+
+    // The timestamp of the last price change
+    uint256 private _priceChangeTimestamp;
+
+    address payable internal _wallet;
+
+    // The delay between calling setPrices() and when the new prices come into effect
+    // Minimizes the effect of price changes on pending TXs
+    uint256 private constant PRICE_CHANGE_DELAY = 1 hours;
+
+    event ReceivingWallet(address newReceivingWallet);
+
+    event Purchase(address indexed buyer, Message message, uint256 amountPaid, address token);
+
+    event SetPrices(
+        uint256[] catalystIds,
+        uint256[] catalystPrices,
+        uint256[] gemIds,
+        uint256[] gemPrices,
+        uint256 priceChangeTimestamp
+    );
+
+    struct Message {
+        uint256[] catalystIds;
+        uint256[] catalystQuantities;
+        uint256[] gemIds;
+        uint256[] gemQuantities;
+        uint256 nonce;
+    }
+
+    constructor(
+        address admin,
+        address sandContractAddress,
+        address trustedForwarder,
+        address payable initialWalletAddress,
+        address initialSigningWallet,
+        address registry
+    ) PurchaseValidator(initialSigningWallet) {
+        _setupRole(DEFAULT_ADMIN_ROLE, admin);
+        _sand = sandContractAddress;
+        __ERC2771Handler_initialize(trustedForwarder);
```

Comment:
weird to use the upgradeable version here since it's not a proxified contract (lets wait for the decision on the proxy to change it)

Reply:
See ERC2771Handler.sol
Would you propose a different approach?